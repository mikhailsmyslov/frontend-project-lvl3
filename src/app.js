import isURL from 'validator/lib/isURL';
import axios from 'axios';
import $ from 'jquery';
import { watch } from 'melanke-watchjs';
import _ from 'lodash';
import {
  proxifyUrl, parseRss, hasUrl, normalizeUrl,
} from './helpers';
import buildItemLi from './components/buildItemLi';
import buildChannelLi from './components/buildChannelLi';

const handleInput = (state, t) => ({ target }) => {
  const { value } = target;
  const urls = state.channels.map(({ url }) => url);
  state.input = value;
  state.errMsgs = [];
  switch (true) {
    case !value:
      state.formState = 'empty';
      break;
    case value && !isURL(value):
      state.formState = 'invalid';
      state.errMsgs.push(t('invalidUrl'));
      break;
    case hasUrl(urls, value):
      state.formState = 'invalid';
      state.errMsgs.push(t('urlAlreadyExist'));
      break;
    default:
      state.formState = 'valid';
  }
};

const runPeriodiсRssUpdate = (url, state, interval, t) => axios
  .get(proxifyUrl(url))
  .then(({ status, data }) => {
    const parsed = parseRss(data);
    if (status === 200 && parsed === null) {
      throw new Error(t('notRss'));
    }

    const normalizedUrl = normalizeUrl(url);
    const newChannel = { ...parsed.channel, url: normalizedUrl };
    const newItems = parsed.items.map(item => ({ ...item, channelUrl: normalizedUrl }));
    const { channels, items } = state;
    const urls = channels.map(c => c.url);

    if (!hasUrl(urls, url)) channels.push(newChannel);
    state.items.unshift(..._.differenceBy(newItems, items, ({ guid }) => guid));

    setTimeout(() => runPeriodiсRssUpdate(url, state, interval, t), interval);
  });

const handleSubmit = (state, period, t) => (event) => {
  event.preventDefault();
  const { target } = event;

  const formData = new FormData(target);
  const url = formData.get('rssUrl');

  state.formState = 'sending';

  runPeriodiсRssUpdate(url, state, period, t)
    .catch((err) => {
      state.errMsgs.push(err.message);
      console.error(err);
    })
    .finally(() => {
      state.formState = _.isEmpty(state.errMsgs) ? 'empty' : 'invalid';
    });
};


export default (translate, period = 5000) => {
  const t = translate || _.identity;
  if (translate) {
    const i18nElements = document.querySelectorAll('*[data-i18n]');
    [...i18nElements].forEach((el) => {
      el.textContent = t(el.dataset.i18n);
    });
  }

  const state = {
    channels: [],
    items: [],
    formState: 'empty',
    errMsgs: [],
    input: '',
  };

  const rssForm = document.getElementById('rssForm');
  const rssInput = rssForm.querySelector('input');
  const formFeedBack = rssForm.querySelector('.feedback');
  const submitButton = rssForm.querySelector('*[type=submit]');
  const channelsUl = document.getElementById('channels');
  const itemsUl = document.getElementById('items');

  rssInput.addEventListener('input', handleInput(state, t));
  rssForm.addEventListener('submit', handleSubmit(state, period, t));

  watch(state, 'formState', () => {
    const { formState, errMsgs } = state;
    switch (formState) {
      case 'empty':
        rssInput.classList.remove('is-invalid', 'is-valid');
        formFeedBack.textContent = '';
        rssForm.reset();
        rssInput.removeAttribute('disabled', '');
        submitButton.setAttribute('disabled', '');
        break;
      case 'valid':
        rssInput.classList.remove('is-invalid');
        rssInput.classList.add('is-valid');
        formFeedBack.textContent = t('formValid');
        rssInput.removeAttribute('disabled', '');
        submitButton.removeAttribute('disabled', '');
        break;
      case 'invalid':
        rssInput.classList.remove('is-valid');
        rssInput.classList.add('is-invalid');
        formFeedBack.textContent = errMsgs.join('. ');
        rssInput.removeAttribute('disabled', '');
        submitButton.setAttribute('disabled', '');
        break;
      case 'sending':
        formFeedBack.textContent = t('loading');
        rssInput.setAttribute('disabled', '');
        submitButton.setAttribute('disabled', '');
        break;
      default:
        break;
    }
  });

  watch(state, 'channels', () => {
    channelsUl.innerHTML = '';
    state.channels.forEach(({ title, description, link }) => {
      const li = buildChannelLi(title, description, link, t);
      channelsUl.appendChild(li);
    });
  });

  watch(state, 'items', () => {
    itemsUl.innerHTML = '';
    state.items.forEach(({ title, description, link }) => {
      const li = buildItemLi(title, description, link, t);
      itemsUl.appendChild(li);
    });
  });

  watch(state, 'input', () => {
    rssInput.value = state.input.toString().trim().toLowerCase();
  });

  $('#modal').on('show.bs.modal', (event) => {
    const button = $(event.relatedTarget);
    const title = button.data('title');
    const body = button.data('body');
    const modal = $(event.target);
    modal.find('.modal-title').text(title);
    modal.find('.modal-body').html(body);
  });
};
