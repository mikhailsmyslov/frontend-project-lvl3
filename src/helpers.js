const proxy = 'https://cors-anywhere.herokuapp.com';
export const proxifyUrl = url => `${proxy}/${url}`;

export const parse = (data) => {
  const parser = new DOMParser();
  const doc = parser.parseFromString(data, 'text/xml');

  const channelDOM = doc.querySelector('channel');
  if (channelDOM === null) return null;

  const channel = channelDOM && {
    title: channelDOM.querySelector('title').textContent,
    description: channelDOM.querySelector('description').textContent,
    link: channelDOM.querySelector('link').textContent,
  };

  const itemsDOM = doc.querySelectorAll('item');
  const items = itemsDOM && [...itemsDOM].map(item => ({
    guid: item.querySelector('guid').textContent,
    title: item.querySelector('title').textContent,
    description: item.querySelector('description').textContent,
    link: item.querySelector('link').textContent,
  }));

  return { channel, items };
};
