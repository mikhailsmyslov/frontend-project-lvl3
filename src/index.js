import 'bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import i18next from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import app from './app';
import en from '../locales/en/translation.json';
import ru from '../locales/ru/translation.json';

i18next
  .use(LanguageDetector)
  .init({
    debug: process.env.NODE_ENV === 'development',
    resources: {
      en: { translation: en },
      ru: { translation: ru },
    },
    fallbackLng: 'en',
  })
  .then((t) => app(t, 5000));
