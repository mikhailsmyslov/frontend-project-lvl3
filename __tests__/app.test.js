import { promises as fs } from 'fs';
import path from 'path';
import { html } from 'js-beautify';
import userEvent from '@testing-library/user-event';
import nock from 'nock';
import { waitForDomChange } from '@testing-library/dom';
import { proxifyUrl } from '../src/helpers';
import app from '../src/app';

const baseUrl = 'http://test.com';
const requestInterval = 2500;

const htmlOptions = {
  preserve_newlines: true,
  unformatted: [],
};

let form;
let input;
let feedback;
let items;

const getTree = () => html(document.body.innerHTML, htmlOptions);

beforeAll(() => {
  nock.disableNetConnect();
  nock(proxifyUrl(baseUrl))
    .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
    .get('/404')
    .reply(404)
    .get('/503')
    .reply(503)
    .get('/rss')
    .replyWithFile(200, path.resolve(__dirname, '__fixtures__/rss1.xml'))
    .get('/rss')
    .replyWithFile(200, path.resolve(__dirname, '__fixtures__/rss2.xml'));
  jest.spyOn(console, 'error').mockImplementation(() => {});
});

beforeEach(async () => {
  const pathToInitHtml = path.resolve(__dirname, '__fixtures__/index.html');
  const initHtml = await fs.readFile(pathToInitHtml, 'utf8');
  document.documentElement.innerHTML = initHtml;

  form = document.getElementById('rssForm');
  input = form.querySelector('input');
  feedback = form.querySelector('.feedback');
  items = document.getElementById('items');

  app(null, requestInterval);
});

test('Should validate URL\'s', async () => {
  input.setAttribute('value', 'invalidUrl');
  await userEvent.type(input, 'invalidUrl', { allAtOnce: true });
  await waitForDomChange({ container: feedback });
  expect(getTree()).toMatchSnapshot();

  input.setAttribute('value', 'valid-url.com');
  await userEvent.type(input, 'valid-url.com', { allAtOnce: true });
  await waitForDomChange({ container: feedback });
  expect(getTree()).toMatchSnapshot();
});

test('Should handle 4xx errors', async () => {
  input.setAttribute('value', `${baseUrl}/404`);
  await form.dispatchEvent(new Event('submit'));
  await waitForDomChange({ container: feedback });
  expect(getTree()).toMatchSnapshot();
});

test('Should handle 5xx errors', async () => {
  input.setAttribute('value', `${baseUrl}/503`);
  await form.dispatchEvent(new Event('submit'));
  await waitForDomChange({ container: feedback });
  expect(getTree()).toMatchSnapshot();
});

test('App', async () => {
  input.setAttribute('value', `${baseUrl}/rss`);
  await form.dispatchEvent(new Event('submit'));
  await waitForDomChange({ container: items });
  expect(getTree()).toMatchSnapshot();

  await waitForDomChange({ container: items });
  expect(getTree()).toMatchSnapshot();
});

test('Should have immutable state', async () => {
  expect(getTree()).toMatchSnapshot();
});
