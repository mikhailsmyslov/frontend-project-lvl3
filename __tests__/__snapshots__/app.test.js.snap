// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 1`] = `
"<form id=\\"rssForm\\">
    <input type=\\"text\\" name=\\"rssUrl\\" required=\\"\\" value=\\"http://test.com/rss\\">
    <button type=\\"submit\\" disabled=\\"\\">Add</button>
    <div class=\\"feedback\\"></div>
</form>
<ul id=\\"channels\\">
    <li class=\\"list-group-item\\"><a href=\\"http://example.com/\\">Lorem ipsum feed for an interval of 1 minutes</a>
        <p>This is a constantly updating lorem ipsum feed</p>
    </li>
</ul>
<ul id=\\"items\\">
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374780\\">Lorem ipsum 2020-01-18T19:13:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:13:00Z\\" data-body=\\"Eiusmod Lorem amet non incididunt.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374720\\">Lorem ipsum 2020-01-18T19:12:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:12:00Z\\" data-body=\\"Cupidatat consectetur cupidatat enim cupidatat ut ad sint duis officia.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374660\\">Lorem ipsum 2020-01-18T19:11:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:11:00Z\\" data-body=\\"Sit in laboris quis eiusmod in commodo do eiusmod cupidatat dolore consectetur.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374600\\">Lorem ipsum 2020-01-18T19:10:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:10:00Z\\" data-body=\\"Id ad est reprehenderit nisi officia consequat laboris cupidatat ut mollit id ipsum duis.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374540\\">Lorem ipsum 2020-01-18T19:09:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:09:00Z\\" data-body=\\"Ad anim in commodo aliqua minim consectetur irure qui.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374480\\">Lorem ipsum 2020-01-18T19:08:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:08:00Z\\" data-body=\\"Veniam sint do duis est ullamco eu voluptate magna ad aute.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374420\\">Lorem ipsum 2020-01-18T19:07:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:07:00Z\\" data-body=\\"Pariatur ad nulla consequat ea irure velit.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374360\\">Lorem ipsum 2020-01-18T19:06:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:06:00Z\\" data-body=\\"Amet sint in sit laborum duis labore qui cillum ad nostrud enim est duis.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374300\\">Lorem ipsum 2020-01-18T19:05:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:05:00Z\\" data-body=\\"Aute sint voluptate nisi excepteur laboris irure enim cillum aliquip esse id sunt consequat nulla.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374240\\">Lorem ipsum 2020-01-18T19:04:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:04:00Z\\" data-body=\\"Cupidatat ad nostrud dolore Lorem deserunt.\\">more</button></li>
</ul>"
`;

exports[`App 2`] = `
"<form id=\\"rssForm\\">
    <input type=\\"text\\" name=\\"rssUrl\\" required=\\"\\" value=\\"http://test.com/rss\\">
    <button type=\\"submit\\" disabled=\\"\\">Add</button>
    <div class=\\"feedback\\"></div>
</form>
<ul id=\\"channels\\">
    <li class=\\"list-group-item\\"><a href=\\"http://example.com/\\">Lorem ipsum feed for an interval of 1 minutes</a>
        <p>This is a constantly updating lorem ipsum feed</p>
    </li>
</ul>
<ul id=\\"items\\">
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579375440\\">Lorem ipsum 2020-01-18T19:24:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:24:00Z\\" data-body=\\"Culpa nisi ea magna ipsum voluptate cupidatat et reprehenderit in labore enim dolore dolore.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579375380\\">Lorem ipsum 2020-01-18T19:23:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:23:00Z\\" data-body=\\"Proident dolore tempor proident aute excepteur id consectetur.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374780\\">Lorem ipsum 2020-01-18T19:13:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:13:00Z\\" data-body=\\"Eiusmod Lorem amet non incididunt.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374720\\">Lorem ipsum 2020-01-18T19:12:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:12:00Z\\" data-body=\\"Cupidatat consectetur cupidatat enim cupidatat ut ad sint duis officia.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374660\\">Lorem ipsum 2020-01-18T19:11:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:11:00Z\\" data-body=\\"Sit in laboris quis eiusmod in commodo do eiusmod cupidatat dolore consectetur.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374600\\">Lorem ipsum 2020-01-18T19:10:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:10:00Z\\" data-body=\\"Id ad est reprehenderit nisi officia consequat laboris cupidatat ut mollit id ipsum duis.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374540\\">Lorem ipsum 2020-01-18T19:09:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:09:00Z\\" data-body=\\"Ad anim in commodo aliqua minim consectetur irure qui.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374480\\">Lorem ipsum 2020-01-18T19:08:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:08:00Z\\" data-body=\\"Veniam sint do duis est ullamco eu voluptate magna ad aute.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374420\\">Lorem ipsum 2020-01-18T19:07:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:07:00Z\\" data-body=\\"Pariatur ad nulla consequat ea irure velit.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374360\\">Lorem ipsum 2020-01-18T19:06:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:06:00Z\\" data-body=\\"Amet sint in sit laborum duis labore qui cillum ad nostrud enim est duis.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374300\\">Lorem ipsum 2020-01-18T19:05:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:05:00Z\\" data-body=\\"Aute sint voluptate nisi excepteur laboris irure enim cillum aliquip esse id sunt consequat nulla.\\">more</button></li>
    <li class=\\"list-group-item d-flex justify-content-between\\"><a href=\\"http://example.com/test/1579374240\\">Lorem ipsum 2020-01-18T19:04:00Z</a><button class=\\"btn btn-primary btn-sm\\" data-toggle=\\"modal\\" data-target=\\"#modal\\" data-title=\\"Lorem ipsum 2020-01-18T19:04:00Z\\" data-body=\\"Cupidatat ad nostrud dolore Lorem deserunt.\\">more</button></li>
</ul>"
`;

exports[`Should handle 4xx errors 1`] = `
"<form id=\\"rssForm\\">
    <input type=\\"text\\" name=\\"rssUrl\\" required=\\"\\" value=\\"http://test.com/404\\" class=\\"is-invalid\\">
    <button type=\\"submit\\" disabled=\\"\\">Add</button>
    <div class=\\"feedback text-danger\\">errors.4xx,errors.oops</div>
</form>
<ul id=\\"channels\\"></ul>
<ul id=\\"items\\"></ul>"
`;

exports[`Should handle 5xx errors 1`] = `
"<form id=\\"rssForm\\">
    <input type=\\"text\\" name=\\"rssUrl\\" required=\\"\\" value=\\"http://test.com/503\\" class=\\"is-invalid\\">
    <button type=\\"submit\\" disabled=\\"\\">Add</button>
    <div class=\\"feedback text-danger\\">errors.5xx,errors.oops</div>
</form>
<ul id=\\"channels\\"></ul>
<ul id=\\"items\\"></ul>"
`;

exports[`Should validate URL's 1`] = `
"<form id=\\"rssForm\\">
    <input type=\\"text\\" name=\\"rssUrl\\" required=\\"\\" value=\\"invalidUrl\\" class=\\"is-invalid\\">
    <button type=\\"submit\\" disabled=\\"\\">Add</button>
    <div class=\\"feedback text-danger\\">errors.invalidUrl,errors.oops</div>
</form>
<ul id=\\"channels\\"></ul>
<ul id=\\"items\\"></ul>"
`;

exports[`Should validate URL's 2`] = `
"<form id=\\"rssForm\\">
    <input type=\\"text\\" name=\\"rssUrl\\" required=\\"\\" value=\\"valid-url.com\\" class=\\"is-valid\\">
    <button type=\\"submit\\">Add</button>
    <div class=\\"feedback text-success\\">formValid</div>
</form>
<ul id=\\"channels\\"></ul>
<ul id=\\"items\\"></ul>"
`;
